enhancements/console/dynamic-plugins.md

```diff
@@ -0,0 +1,389 @@
+* Plugins won't be sandboxed. They will have full JavaScript access to the DOM and network.
+* This proposal does not cover allowing plugins to contribute backend console endpoints.
+
+## Proposal
+
+### Module Federation
+
+Console will use [Webpack module federation](https://webpack.js.org/concepts/module-federation/)
+to load plugins at runtime. Module federation allows a JavaScript application
+to dynamically load code from another application while sharing dependencies.
+It also allows console to share its components with plugins built and bundled
+separately.
+
+Plugins will need to be built with Webpack 5 to use module federation.
+
+We have a pull request with a [working prototype](https://github.com/openshift/console/pull/6101)
+that uses module federation.
+
+### Shared Dependencies
+
+Console will expose React and Patternfly 4 as shared dependencies to plugins.
+Plugins can use any Patternfly component. Only a single version of React and
+Patternfly will be loaded.
+
+Plugins will not be able to use legacy Patternfly 3 components.
+
+### Delivering Plugins
+
+Plugins are delivered through operators. The operator will create a deployment
+on the platform with an HTTP server to host the plugin and expose it using a
+Kubernetes `Service`. The HTTP server serves all assets for the plugin,
+including JavaScript, CSS, and images. The `Service` must use HTTPS and a
+[service serving certificate](https://docs.openshift.com/container-platform/4.4/security/certificates/service-serving-certificate.html).
+The console backend will proxy the plugin assets from the `Service` using the
+service CA bundle.
+
+Operators declare that they have a console plugin available by creating a
+cluster-scoped `ConsolePlugin` resource that includes the service name, port,
+and path to the plugin manifest.
+
+```yaml
+apiVersion: console.openshift.io/v1alpha1
+kind: ConsolePlugin
+metadata:
+  name: my-plugin
+spec:
+  displayName: My Plugin
+  service:
+    name: my-console-plugin
+    namespace: my-operator-namespace
+    port: 8443
+    manifest: /manifest.json
+```
+
+Plugins are disabled by default. They need to be manually enabled by a cluster
+administrator before console loads any plugin code. Console provides a UI
+for enabling plugins, and the list of enabled plugins is set in the console
+operator config.
+
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: Console
+metadata:
+  name: cluster
+spec:
+  managementState: Managed
+  plugins:
+  - my-plugin
+```
+
+This will trigger a new rollout of the console `Deployment` with the updated
+plugins. Note that this does mean the user will need to refresh the browser
+to see the new plugins.
+
+An operator can also give console a hint that it has a plugin available through
+the `console.openshift.io/plugins` annotations. The annotation tells the console to
+show a checkbox to enable plugins during operator install. The value is a
```

Comment:
This kinda contradicts the previous statement about enabling the plugins
> They need to be manually enabled by a cluster administrator before console loads any plugin code...

Reply:
I don't think it does? We provide a UI to do it, but an admin still needs to do something to enable the plugin.

Reply:
It just confused me a bit since:
```
The annotation tells the console to show a checkbox to enable plugins during operator install.
```
thought this would enable the plugin, just if the user checks the checkbox (without admin being needed). 
Will admin need to enable it anyway?

Reply:
Only admins will see the checkbox. (Typically, only admins can install operators.) I can clarify in the doc.